### ISSUE TITLE
Enhance DataGridView to support non-destructive (multi column) Sort

### ISSUE THREAD

**Issue by coderb** [dotnet/winforms#5202]
`DataGridView.Sort` seems to be implemented via the quick sort algo in `DataGridViewRowCollection.RowArrayList`.

I would like to be able to sort the DataGridView in a non-destructive (stable) way when changing the sort column (clicking the header to sort).  Eg.  If I sort on the first column and then sort on the second, the rows remain sorted by column 1 for equal values of column 2.

**Comment by coderb** [comment]
@krwq nope, that link is a no go.  i need to be able to (continue to) use the column headers to do the sorting.  i've tried everything i could find on the web/stack overflow and even tried doing some reflection to SortInternal when overriding the Sort() method of DataGridView but it doesn't like to be passed both a column and comparer object.

it seems like some code support in DataGridView is appropriate.

**Comment by coderb** [comment]
i've requested that the dotnet core project add a stable sort algorithm. 
https://github.com/dotnet/core/issues/7342

**Comment by jkotas** [comment]
@corerb `DataGridViewRowCollection.RowArrayList` has its own custom sort implementation:  https://github.com/dotnet/winforms/blob/d46ad2e8dc76248739d9ae22b28b399a6a6b299e/src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowCollection.RowArrayList.cs#L35 . Unless the implementation is refactored to use the sort implementation from the core libraries, you would not be able to use stable sort even if it was added to the core libraries.

Also, it should be fairly straightforward to modify the compare function to take the current order into account. It may take a bit extra memory (e.g. temporary array of integers). It is negliable given how much memory is occupied by each row already.

**Comment by JeremyKuhne** [comment]
@KlausLoeffelmann this sounds like a useful feature, thoughts?

**Comment by KlausLoeffelmann** [comment]
Hmm. I am a bit torn, also since Data Binding might be an issue here, since - if I remember correctly - sorting will then be delegated to a `BindingList<T>` and not be done by the `DataGridView`? But - I would need to refresh my memory what exactly is happening.

For unbound data sources, I would think implementing a `MultiColumnSorter` (based on `IComparer`) to pass to a DataGridView's `Sort` method should work already.

So, I don't know. Am I missing something else which should justify this feature?

**Comment by KlausLoeffelmann** [comment]
I am moving this to future. We won't have the bandwidth in the nearer future to address this.
